mlflow/fastai/callback.py

```diff
@@ -0,0 +1,157 @@
+from mlflow.fastai import log_model
+
+from fastai.callback.core import Callback
+
+
+# Move outside, because it cannot be pickled. Besides, ExceptionSafeClass was giving some issues
+class __MLflowFastaiCallback(Callback):
+    """
+    Callback for auto-logging metrics and parameters.
+    Records model structural information as params when training begins.
+    """
+
+    from fastai.learner import Recorder
+    from fastai.callback.all import TrackerCallback
+
+    remove_on_fetch, run_before, run_after = True, TrackerCallback, Recorder
+
+    def __init__(self, metrics_logger, log_models, is_fine_tune=False):
+        super().__init__()
+        self.metrics_logger = metrics_logger
+        self.log_models = log_models
+        self.is_fine_tune = is_fine_tune
+        self.freeze_prefix = ""
+
+    def after_epoch(self):
+        """Log loss and other metrics values after each epoch"""
+
+        def _is_float(x):
+            try:
+                float(x)
+                return True
+            except (ValueError, TypeError):
+                return False
+
+        # Do not record in case of predicting
+        if hasattr(self, "lr_finder") or hasattr(self, "gather_preds"):
+            return
+        # Remove non-float metrics and record the rest.
+        metrics = self.recorder.log
+        metrics = dict(zip(self.recorder.metric_names, metrics))
+        for key in [key for key, val in metrics.items() if not _is_float(val)]:
+            del metrics[key]
+        self.metrics_logger.record_metrics(metrics, step=metrics["epoch"])
+
+    def before_fit(self):
+        from fastai.callback.all import ParamScheduler
+
+        # Do not record in case of predicting or lr_finder
+        if hasattr(self, "lr_finder") or hasattr(self, "gather_preds"):
+            return
+
+        if self.is_fine_tune and len(self.opt.param_lists) == 1:
+            print(
+                """ WARNING: Using `fine_tune` with model which cannot be frozen.
+                Current model have only one param group which makes it impossible to freeze.
+                Because of this it will record some fitting params twice (overriding exception)
+                """
+            )
+
+        frozen = self.opt.frozen_idx != 0
+        if frozen and self.is_fine_tune:
+            self.freeze_prefix = "freeze_"
+        else:
+            self.freeze_prefix = ""
+
+        # Extract function name when `opt_func` is partial function
+        if isinstance(self.opt_func, partial):
+            try_mlflow_log(
+                mlflow.log_param,
+                self.freeze_prefix + "opt_func",
+                self.opt_func.keywords["opt"].__name__,
+            )
+        else:
+            try_mlflow_log(
+                mlflow.log_param, self.freeze_prefix + "opt_func", self.opt_func.__name__
+            )
+
+        params_not_to_log = []
+        for cb in self.cbs:
+            if isinstance(cb, ParamScheduler):
```

Comment:
For param schedulers, can we also log the schedular name and its start& end values?

```
sched = {'lr': SchedLin(start=1e-3, end=1e-2)}
learn.fit(1, cbs=ParamScheduler(sched))
```

Reply:
Ah the start and end values are logged as `<param>_init` and `<param>_final`